libyul/optimiser/StackLimitEvader.cpp

```diff
@@ -0,0 +1,140 @@
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/StackLimitEvader.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/FunctionCallFinder.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+// Walks the call graph using a Depth-First-Search assigning memory offsets to variables.
+// - The leaves of the call graph will get the lowest offsets, increasing towards the root.
+// - ``nextAvailableSlot`` maps a function to the next available slot that can be used by another
+//   function that calls it.
+// - For each function starting from the root of the call graph:
+//   - Visit all children that are not already visited.
+//   - Determe the maximum value ``n`` of the values of ``nextAvailableSlot`` among the children.
+//   - If the function itself contains variables that need memory slots, but is contained in a cycle,
+//     abort the process as failure.
+//   - If not, assign each variable its slot starting starting from ``n`` (incrementing it).
+//   - Assign ``nextAvailableSlot`` of the function to ``n``.
+struct MemoryOffsetAllocator
+{
+	map<YulString, set<YulString>> const& unreachableVariables;
+	map<YulString, set<YulString>> const& callGraph;
+
+	uint64_t run(YulString _function = YulString{})
+	{
+		if (nextAvailableSlot.count(_function))
+			return nextAvailableSlot[_function];
+
+		// Assign to zero early to guard against recursive calls.
+		nextAvailableSlot[_function] = 0;
+
+		uint64_t nextSlot = 0;
+		if (callGraph.count(_function))
+			for (YulString child: callGraph.at(_function))
+				nextSlot = std::max(run(child), nextSlot);
+
+		if (unreachableVariables.count(_function))
+		{
+			yulAssert(!slotAllocations.count(_function), "");
+			auto& assignedSlots = slotAllocations[_function];
+			for (YulString variable: unreachableVariables.at(_function))
+				if (variable.empty())
+				{
+					// TODO: Too many function arguments or return parameters.
+				}
+				else
+					assignedSlots[variable] = nextSlot++;
+		}
+
+		return (nextAvailableSlot[_function] = nextSlot);
+	}
+
+	map<YulString, map<YulString, uint64_t>> slotAllocations{};
+	map<YulString, uint64_t> nextAvailableSlot{};
+};
+u256 getLiteralArgumentValue(FunctionCall const& _call)
+{
+	yulAssert(_call.arguments.size() == 1, "");
+	Literal const* literal = std::get_if<Literal>(&_call.arguments.front());
```

Comment:
Hm, should we also assert `literal.kind == LiteraelKind::Number`? The `valueOfLiteral()` implicitly converts bools/strings.

Reply:
Especially since it implicitly converts bools and strings (i.e. this will never actually break), I actually *don't* see a need to assert this again here... but well, I can still add it.

Reply:
What I'm worried about having implicitly converted bools/strings here to be honest.

Reply:
Well, we could argue about that, but I added it anyways.